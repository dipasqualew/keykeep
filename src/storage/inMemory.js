/**
 * @module Storage/InMemoryHandler
 */


import StorageHandler from './handler';

/**
 * In Memory Handler
 *
 * @constructor
 * @extends Storage:StorageHandler
 */
export default class InMemoryHandler extends StorageHandler {

    constructor(key, di) {
        super(key, di);
        this.__storage__ = null;
    }

    /** @inheritdoc */
    async implSave(_context, data) {
        this.__storage__ = data;
    }

    /**
     * @override
     */
    implLoad() {
        return this.__storage__;
    }

    /**
     * @inheritdoc
     */
    implClear() {
        return this.__storage__ = null;
    }

}
