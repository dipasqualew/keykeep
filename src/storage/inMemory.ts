/**
 * @module Storage/InMemoryHandler
 */

import { EncryptionContext } from '../strategy/handler';
import {
  StorageHandler,
  StorageContext,
  Serialized,
} from './handler';


/**
 * In Memory Handler
 */
export class InMemoryHandler extends StorageHandler {

  STORAGE: Serialized | null;
  static serializer = 'InMemoryHandler';

  constructor(key: string, di: StorageContext) {
    super(key, di);
    this.STORAGE = null;
  }

  async implSave(_context: EncryptionContext, serialized: Serialized): Promise<void> {
    this.STORAGE = serialized;
  }

  async implLoad(_context: EncryptionContext): Promise<Serialized | null> {
    return this.STORAGE;
  }

  async implClear(_context: EncryptionContext): Promise<void> {
    this.STORAGE = null;
  }

}
