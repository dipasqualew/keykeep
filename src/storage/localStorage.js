
import StorageHandler from './handler';

/**
 * LocalStorage Handler
 */
export default class LocalStorageHandler extends StorageHandler {

    /**
     * Returns the LocalStorage instance.
     *
     * @returns {WindowLocalStorage}
     */
    get ls() {
        return this.context.localStorage;
    }

    /**
     * @inheritdoc
     */
    async implSave(context, data) {
        this.ls.setItem(this.key, data);
    }

    /**
     * @inheritdoc
     */
    implLoad() {
        return this.ls.getItem(this.key);
    }

}
