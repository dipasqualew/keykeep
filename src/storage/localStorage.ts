/**
 * @module Storage/LocalStorageHandler
 */


import { EncryptionContext } from '../strategy/handler';
import { StorageHandler, Serialized } from './handler';


/**
 * LocalStorage Handler
 *
 * @constructor
 */
export class LocalStorageHandler extends StorageHandler {
  static serializer = 'LocalStorageHandler';

  get ls(): Storage {
    return window.localStorage;
  }

  async implSave(_context: EncryptionContext, serialized: Serialized): Promise<void> {
    this.ls.setItem(this.key, JSON.stringify(serialized));
  }

  async implLoad(_context: EncryptionContext): Promise<Serialized | null> {
    const raw = this.ls.getItem(this.key);

    if (raw) {
      return JSON.parse(raw);
    }

    return null;
  }

  async implClear(_context: EncryptionContext): Promise<void> {
    this.ls.removeItem(this.key);
  }

}
