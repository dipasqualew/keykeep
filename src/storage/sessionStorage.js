/**
 * @module Storage/SessionStorageHandler
 */


import StorageHandler from './handler';

/**
 * SessionStorage Handler
 *
 * @constructor
 */
export default class SessionStorageHandler extends StorageHandler {

    /**
     * Returns the SessionStorage instance.
     *
     * @returns {WindowSessionStorage}
     */
    get ss() {
        return this.context.sessionStorage;
    }

    /**
     * @inheritdoc
     */
    async implSave(context, data) {
        this.ss.setItem(this.key, JSON.stringify(data));
    }

    /**
     * @inheritdoc
     */
    implLoad() {
        return JSON.parse(this.ss.getItem(this.key));
    }

    /**
     * @inheritdoc
     */
    async implClear() {
        this.ss.removeItem(this.key);
    }

}
