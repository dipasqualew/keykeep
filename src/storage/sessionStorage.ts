/**
 * @module Storage/SessionStorageHandler
 */


import { EncryptionContext } from '../strategy/handler';
import { StorageHandler, Serialized } from './handler';

/**
 * SessionStorage Handler
 */
export class SessionStorageHandler extends StorageHandler {

  static serializer = 'SessionStorageHandler';

  /**
     * Returns the SessionStorage instance.
     */
  get ss(): Storage {
    return window.sessionStorage;
  }

  async implSave(_context: EncryptionContext, serialized: Serialized): Promise<void> {
    const raw = JSON.stringify(serialized);
    this.ss.setItem(this.key, raw);
  }

  async implLoad(_context: EncryptionContext): Promise<Serialized | null> {
    const raw = this.ss.getItem(this.key);

    if (raw) {
      return JSON.parse(raw);
    }

    return null;
  }

  async implClear(_context: EncryptionContext): Promise<void> {
    this.ss.removeItem(this.key);
  }

}
