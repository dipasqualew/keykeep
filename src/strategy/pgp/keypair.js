/**
 * @module Strategy
 */


import { message, decrypt, encrypt, key } from 'openpgp';

import StrategyHandler from '../handler';

/**
 * PGP KeypairHandler Handler
 *
 * @constructor
 */
export default class KeypairHandler extends StrategyHandler {

    /**
     * Read and decrypts
     * public and private keys
     *
     * @param {object} context
     * @param {string} context.publicKey
     * @param {string} context.privateKey
     *
     * @returns {object}
     */
    async readKeys(context) {
        const [{ keys: publicKeys }, { keys: privateKeys }] = await Promise.all([
            key.readArmored(context.publicKey),
            key.readArmored(context.privateKey),
        ]);

        await Promise.all(privateKeys.map((pk) => pk.decrypt(context.passphrase)));

        return {
            publicKeys,
            privateKeys,
        };
    }

    /**
     * Encrypts a string with the given password.
     *
     * @param {object} context
     * @param {string} context.password
     * @param {string} plaintext
     * @returns {string}
     */
    async encrypt(context, plaintext) {
        const { publicKeys, privateKeys } = await this.readKeys(context);

        const options = {
            message: message.fromText(plaintext),
            publicKeys,
            privateKeys,
            armor: true,
        };

        const output = await encrypt(options);

        return output.data;
    }

    /**
     * Decrypts a string with the given password.
     *
     * @param {object} context
     * @param {string} context.password
     * @param {string} encrypted
     * @returns {string}
     */
    async decrypt(context, encrypted) {
        const { publicKeys, privateKeys } = await this.readKeys(context);

        const options = {
            message: await message.readArmored(encrypted),
            publicKeys,
            privateKeys,
        };

        const output = await decrypt(options);

        return output.data;
    }
}
