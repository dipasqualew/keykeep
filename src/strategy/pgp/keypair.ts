/**
 * @module Strategy
 */


import {
  message,
  decrypt,
  encrypt,
  key,
  DecryptOptions,
  EncryptOptions,
} from 'openpgp';

import { StrategyHandler, EncryptionContext } from '../handler';

interface KeypairEncryptionContext extends EncryptionContext {
    publicKey: string,
    privateKey: string,
    passphrase: string,
}

interface DecryptedKeypair {
    publicKeys: key.Key[];
    privateKeys: key.Key[];
}

/**
 * PGP KeypairHandler Handler
 */
export class KeypairHandler extends StrategyHandler<KeypairEncryptionContext> {

  /**
     * Read and decrypts
     * public and private keys
     */
  async readKeys(context: KeypairEncryptionContext): Promise<DecryptedKeypair> {
    const [{ keys: publicKeys }, { keys: privateKeys }] = await Promise.all([
      key.readArmored(context.publicKey),
      key.readArmored(context.privateKey),
    ]);

    await Promise.all(privateKeys.map((pk) => pk.decrypt(context.passphrase)));

    return {
      publicKeys,
      privateKeys,
    };
  }

  /**
     * Encrypts a string with the given password.
     */
  async encrypt(context: KeypairEncryptionContext, plaintext: string): Promise<string> {
    const { publicKeys, privateKeys } = await this.readKeys(context);

    const options: EncryptOptions = {
      message: message.fromText(plaintext),
      publicKeys,
      privateKeys,
      armor: true,
    };

    const output = await encrypt(options);

    return output.data;
  }

  /**
     * Decrypts a string with the given password.
     */
  async decrypt(context: KeypairEncryptionContext, encrypted: string): Promise<string> {
    const { publicKeys, privateKeys } = await this.readKeys(context);

    const options: DecryptOptions = {
      message: await message.readArmored(encrypted),
      publicKeys,
      privateKeys,
    };

    const output = await decrypt(options);

    return output.data;
  }
}
