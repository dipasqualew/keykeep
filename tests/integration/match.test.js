import sinon from 'sinon';

import TextEncoding from 'text-encoding-utf-8';

import { InMemoryHandler } from '../../src/storage/inMemory';
import { LocalStorageHandler } from '../../src/storage/localStorage';
import { SessionStorageHandler } from '../../src/storage/sessionStorage';
import { KeypairHandler } from '../../src/strategy/pgp/keypair';
import { PasswordHandler } from '../../src/strategy/pgp/password';
import {
  getLocalStorage,
  getSessionStorage,
  PAYLOAD,
  PASSWORD,
  PUBLIC_KEY,
  PRIVATE_KEY,
  KEYPAIR_PASSPHRASE,
  ENCRYPTED_WITH_PASSWORD,
  ENCRYPTED_WITH_KEYPAIR,
} from '../mocks';


/**
 * Context to be passed
 * to password-based strategies.
 *
 * @const {object}
 */
const PASSWORD_CONTEXT = {
  password: PASSWORD,
  encryptedBlock: ENCRYPTED_WITH_PASSWORD,
};

/**
 * Context to be passed
 * to keypair-based strategies.
 *
 * @const {object}
 */
const KEYPAIR_CONTEXT = {
  publicKey: PUBLIC_KEY,
  privateKey: PRIVATE_KEY,
  passphrase: KEYPAIR_PASSPHRASE,
  encryptedBlock: ENCRYPTED_WITH_KEYPAIR,
};

/**
 * Runs the integration tests
 * for a combination of strategy and storage.
 *
 * @param {Function} getStrategy
 * @param {Function} getStorage
 * @param {object} context
 */
const integrationMatchTest = (getStrategy, getStorage, context) => {
  beforeAll(() => {
    global.TextEncoder = TextEncoding.TextEncoder;

  });

  const strategyName = getStrategy().constructor.name;
  const storageName = getStorage().constructor.name;

  describe(`${storageName} + ${strategyName}`, () => {

    it('Encrypts a string', async () => {
      const strategy =  getStrategy();
      const storage = getStorage(strategy);
      const encrypted = await storage.serialize(context, PAYLOAD);

      expect(encrypted).toMatchSnapshot({
        meta: {
          datetime: expect.any(Number),
        },
        payload: expect.any(String),
      });

      expect(encrypted.payload).toContain('-----BEGIN PGP MESSAGE-----');
      expect(encrypted.payload).toContain('-----END PGP MESSAGE-----');
    });

    it('Decrypts a string', async () => {
      const strategy = getStrategy();
      const storage = getStorage(strategy);
      const serialized = {
        meta: {
          serializer: storage.constructor.name,
          version: storage.constructor.version,
        },
        payload: context.encryptedBlock,
      };

      const decrypted = await storage.deserialize(context, serialized);

      expect(decrypted).toEqual(PAYLOAD);
    });

    it('Encrypts and decrypts the payload', async () => {
      const strategy =  getStrategy();
      const storage = getStorage(strategy);


      await storage.save(context, PAYLOAD);
      const actual = await storage.load(context);

      return expect(PAYLOAD).toEqual(actual);
    });

    it('Clears the storage', async () => {
      const strategy = getStrategy();
      const storage = getStorage(strategy);

      await storage.save(context, PAYLOAD);
      await storage.clear();

      const decrypted = await storage.load(context);

      expect(decrypted).toEqual(null);
    });
  });
};

/**
 * Returns a configured PasswordHandler.
 *
 * @returns {PasswordHandler}
 */
const getPasswordHandler = () => new PasswordHandler();

/**
 * Returns a configured KeypairHandler.
 *
 * @returns {KeypairHandler}
 */
const getKeypairHandler = () => new KeypairHandler();

/**
 * Returns a configured LocalStorageHandler.
 *
 * @param {StrategyHandler} strategy
 * @returns {LocalStorageHandler}
 */
const getLocalStorageHandler = (strategy) => {
  const ls = getLocalStorage();
  const lsh = new LocalStorageHandler('lsh', { strategy });
  sinon.stub(lsh, 'ls').get(() => ls);

  return lsh;
};

/**
 * Returns a configured SessionStorageHandler.
 *
 * @param {StrategyHandler} strategy
 * @returns {SessionStorageHandler}
 */
const getSessionStorageHandler = (strategy) => {
  const ss = getSessionStorage();
  const ssh = new SessionStorageHandler('ssh', { strategy });
  sinon.stub(ssh, 'ss').get(() => ss);

  return ssh;
};

/**
 * Returns a configured InMemoryHandler.
 *
 * @param {StrategyHandler} strategy
 * @returns {InMemoryHandler}
 */

const getInMemoryHandler = (strategy) => new InMemoryHandler('imh', { strategy });

const strategies = [
  [getPasswordHandler, PASSWORD_CONTEXT],
  [getKeypairHandler, KEYPAIR_CONTEXT],
];

const storages = [
  getInMemoryHandler,
  getLocalStorageHandler,
  getSessionStorageHandler,
];

strategies.forEach(([getStrategy, context]) => {
  storages.forEach((getStorage) => {
    integrationMatchTest(getStrategy, getStorage, context);
  });
});
