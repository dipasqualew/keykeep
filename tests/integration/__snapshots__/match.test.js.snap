// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InMemoryHandler + KeypairHandler Encrypts a string 1`] = `
Object {
  "meta": Object {
    "datetime": Any<Number>,
    "key": "imh",
    "serializer": "InMemoryHandler",
    "version": 1,
  },
  "payload": Any<String>,
}
`;

exports[`InMemoryHandler + PasswordHandler Encrypts a string 1`] = `
Object {
  "meta": Object {
    "datetime": Any<Number>,
    "key": "imh",
    "serializer": "InMemoryHandler",
    "version": 1,
  },
  "payload": Any<String>,
}
`;

exports[`LocalStorageHandler + KeypairHandler Encrypts a string 1`] = `
Object {
  "meta": Object {
    "datetime": Any<Number>,
    "key": "lsh",
    "serializer": "LocalStorageHandler",
    "version": 1,
  },
  "payload": Any<String>,
}
`;

exports[`LocalStorageHandler + PasswordHandler Encrypts a string 1`] = `
Object {
  "meta": Object {
    "datetime": Any<Number>,
    "key": "lsh",
    "serializer": "LocalStorageHandler",
    "version": 1,
  },
  "payload": Any<String>,
}
`;

exports[`SessionStorageHandler + KeypairHandler Encrypts a string 1`] = `
Object {
  "meta": Object {
    "datetime": Any<Number>,
    "key": "ssh",
    "serializer": "SessionStorageHandler",
    "version": 1,
  },
  "payload": Any<String>,
}
`;

exports[`SessionStorageHandler + PasswordHandler Encrypts a string 1`] = `
Object {
  "meta": Object {
    "datetime": Any<Number>,
    "key": "ssh",
    "serializer": "SessionStorageHandler",
    "version": 1,
  },
  "payload": Any<String>,
}
`;
